src/n64sys.c

```diff
@@ -175,6 +180,41 @@ void wait_ms( unsigned long wait_ms )
     wait_ticks(TICKS_FROM_MS(wait_ms));
 }
 
+void sys_rcp_halt(void)
+{
+    // Can't have any interrupts here
+    disable_interrupts();
+
+    // Stop PI transaction
+    *PI_STATUS = PI_STATUS_CLEAR_INTERRUPT | PI_STATUS_RESET;
+
+    // Halt the RSP
+    *SP_STATUS = SP_WSTATUS_SET_HALT;
+
+    // Wait for RSP to halt execution
+    while( !(*SP_STATUS & SP_STATUS_HALTED) );
+
+    // Wait for RSP IO and DMA operations to finish
+    while( *SP_STATUS & (SP_STATUS_IO_BUSY | SP_STATUS_DMA_BUSY) );
+
+    // Flush the RDP and wait for DMA to finish fetching commands
+    *DP_STATUS = DP_WSTATUS_SET_FLUSH;
+    RSP_WAIT_LOOP(200) {
+        if ( !(*DP_STATUS & DP_STATUS_DMA_BUSY) ) {
+            break;
+        }
+    }
+    *DP_STATUS = DP_WSTATUS_RESET_FLUSH;
```

Comment:
Currently this code does not safely shutdown the RDP - there are occasional cases of RDP hard lock that prevents console from using it until full power cycle, thus limiting the soft reboot functionality severely.